zingo-sync/src/scan/compact_blocks/runners.rs

```diff
@@ -306,47 +290,108 @@ pub(crate) struct Batch<D: BatchDomain, Output, Dec: Decryptor<D, Output>> {
 {
     fn dynamic_usage(&self) -> usize {
-        self.tags.dynamic_usage()
-            + self.ivks.dynamic_usage()
-            + self.outputs.dynamic_usage()
-            + self.repliers.dynamic_usage()
+        self.tags.dynamic_usage() + self.ivks.dynamic_usage() + self.outputs.dynamic_usage()
     }
 
     fn dynamic_usage_bounds(&self) -> (usize, Option<usize>) {
         let (tags_lower, tags_upper) = self.tags.dynamic_usage_bounds();
         let (ivks_lower, ivks_upper) = self.ivks.dynamic_usage_bounds();
         let (outputs_lower, outputs_upper) = self.outputs.dynamic_usage_bounds();
-        let (repliers_lower, repliers_upper) = self.repliers.dynamic_usage_bounds();
 
         (
-            tags_lower + ivks_lower + outputs_lower + repliers_lower,
-            tags_upper
-                .zip(ivks_upper)
-                .zip(outputs_upper)
-                .zip(repliers_upper)
-                .map(|(((a, b), c), d)| a + b + c + d),
+            tags_lower + ivks_lower + outputs_lower,
+            // The following is more concise, but harder to read
+            // [tags_upper, ivks_upper, outputs_upper]
+            //    .into_iter().try_fold(0, |a, b| Some(a, b?))
+            match (tags_upper, ivks_upper, outputs_upper) {
+                (Some(a), Some(b), Some(c)) => Some(a + b + c),
+                _ => None,
+            },
         )
     }
 }
 
-impl<D, Output, Dec> Batch<D, Output, Dec>
+pub(crate) trait Batch<D>: Task + Sized
 where
-    D: BatchDomain,
-    Dec: Decryptor<D, Output>,
+    D: BatchDomain<IncomingViewingKey: Send, Recipient: Send, Memo: Send>,
 {
-    /// Constructs a new batch.
-    fn new(tags: Vec<KeyId>, ivks: Vec<D::IncomingViewingKey>) -> Self {
+    /// The data needed to create the batch
+    type Initial;
+
+    /// The items to be processed, in the batch
+    type Input;
+
+    /// The result of processing the items
+    type ResultVal: Send;
+
+    /// As we may return more than one result, we want a unique
+    /// identifier for each
+    type ResultKey: Hash + Eq;
+
+    /// The key used to identify a batch
+    type BatchKey: Hash + Eq + DynamicUsage;
+
+    fn new(init: Self::Initial) -> Self;
+
+    fn inputs(&self) -> &Vec<Self::Input>;
+    fn inputs_mut(&mut self) -> &mut Vec<Self::Input>;
+
+    fn repliers_mut(&mut self) -> &mut Vec<(usize, channel::Sender<(usize, Self::ResultVal)>)>;
+
+    fn is_empty(&self) -> bool {
+        self.inputs().is_empty()
+    }
+
+    /// Adds the given inputs to this batch.
+    ///
+    /// `replier` will be called with the result of every output.
+    fn add_widgets(
+        &mut self,
+        widgets: impl ExactSizeIterator<Item = Self::Input>,
+        replier: channel::Sender<(usize, Self::ResultVal)>,
+    ) {
+        let widget_len = widgets.len();
+        self.inputs_mut().extend(widgets);
+        self.repliers_mut()
+            .extend((0..widget_len).map(|output_index| (output_index, replier.clone())));
```

Comment:
i am making an assumption here but i imagine there is a distinct difference between trial decrpytion and shardtree where trial decrypting and returning a result for each individual output is neccessary but for shardtree we would probably have a result for the insertion of the whole batch? would we need a replier for each node? it likely this is just easier for now, maybe an optimisation we can work on later

Reply:
Agreed! The whole batch should resolve to a single `LocatedPrunableTree`, the exact details will depend on how the implementation of the `ShardTreeUpdateBatch` shakes out